FontIcon markup extension

Sergio0694 opened the pull request:
This PR is part of the group being tracked in #3108.

Also includes changes originally from #3129 and #3130.

## PR Type
What kind of change does this PR introduce?
<!-- Please uncomment one ore more that apply to this PR -->

<!-- - Bugfix -->
 - Feature
<!-- - Code style update (formatting) -->
<!-- - Refactoring (no functional changes, no api changes) -->
<!-- - Build or CI related changes -->
<!-- - Documentation content changes -->
<!-- - Sample app changes -->
<!-- - Other... Please describe: -->

## What is the new behavior?
Some XAML markup extensions to create `FontIcon`-s, `FontIconSource`-s and `BitmapIcon`-s directly from XAML with no code behind, resulting in much cleaner code ([see here](https://user-images.githubusercontent.com/10199417/73084286-fac9c880-3ecc-11ea-8b7c-ec36430a37e9.png)).
All additional info can be found in the linked issue.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] Tested code with current [supported SDKs](../readme.md#supported)
- [X] Pull Request has been submitted to the documentation repository [instructions](..\contributing.md#docs). Link: [\#313](https://github.com/MicrosoftDocs/WindowsCommunityToolkitDocs/pull/313)
- [X] Sample in sample app has been added / updated (for bug fixes / features)
    - [ ] ~Icon has been created (if new sample) following the [Thumbnail Style Guide and templates](https://github.com/windows-toolkit/WindowsCommunityToolkit-design-assets)~
- [X] Tests for the changes have been added (for bug fixes / features) (if applicable)
- [X] Header has been added to all new source files (run *build/UpdateHeaders.bat*)
- [X] Contains **NO** breaking changes


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. 
     Please note that breaking changes are likely to be rejected -->


## Other information

I'm going to need some help in setting the tests up, since they keep failing on my end (despite the fact that the extension in question has been used by a number of apps for years), and I'm wondering whether there are additional steps I missed in cloning/setting up the repository.
If anyone has a minute to double check, I'd appreciate it! 😊

Because of this, I'm only marking this PR as a draft for now.

**EDIT:** seems like the tests are actually passing in the CI. At this point I really think I'm just missing some configuration steps to get them to work on my end as well 😆

Sergio0694 commented:
I'm wondering whether it'd be a good idea to also add relaying properties to the other available [`FontIcon`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.fonticon) properties as , if we want APIs in the toolkit to be as flexible as possible.
In particular, those would be:

- `FontSize`
- `FontStyle`
- `FontWeight`
- `IsTextScaleFactorEnabled`
- `MirroredWhenRightToLeft`

Also, while I'm at it, I have 2 more equivalent markup extensions just like this one, but for [`FontIconSource`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.fonticonsource) (which is common in controls like the WinUI `SwipeItems`), and for [`BitmapIcon`](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.bitmapicon).

I personally think that having markup extensions for these 3 icon types makes it much easier to set things up in XAML, and it also makes the resulting code cleaner and easier to read.

@michael-hawker Should I open two PRs for those other 2 markup extensions as well? 😄

michael-hawker commented:
@Sergio0694 exposing more properties probably isn't a bad idea, if anything `FontSize` is the most important.

Feel free to submit PRs for the other ones as well. you could add the FontIconSource to this one, or just create separate smaller PRs (easier to review that way). :)

michael-hawker commented:
@Sergio0694 just to clarify our UnitTests don't run in our CI yet, I should open an issue to track that.  I can try and pull this down and see if I can run the tests.

michael-hawker commented:
@Sergio0694 I was able to run your tests locally. Think you had some copy-paste bugs.  I didn't notice them on the first look too.  I'll point them out as GitHub suggestions.

michael-hawker commented on UnitTests/Extensions/Test_FontIconExtensionMarkupExtension.cs:
```diff
@@ -0,0 +1,63 @@
+    xmlns=""http://schemas.microsoft.com/winfx/2006/xaml/presentation""
+    xmlns:x=""http://schemas.microsoft.com/winfx/2006/xaml""
+    xmlns:ex=""using:Microsoft.Toolkit.Uwp.UI.Extensions"">
+        <AppBarButton x:Name=""Check"" Icon=""{ex:FontIcon Glyph=&#xE14D;}""/>
```
```suggestion
        <AppBarButton x:Name=""Check"" Icon=""{ex:FontIcon Glyph=&#xE105;}""/>
```
You accidently weren't checking for the same char! 🤦‍♂️ 😋

Sergio0694 replied:
Whoops, thanks! 😄

michael-hawker commented on UnitTests/Extensions/Test_FontIconExtensionMarkupExtension.cs:
```diff
@@ -0,0 +1,63 @@
+
+            Assert.IsNotNull(icon, $"Could not find the {nameof(FontIcon)} element in button.");
+
+            Assert.AreEqual(icon.Glyph, "\uE105", "Expected icon glyph to be E105.");
```
```suggestion
            Assert.AreEqual(icon.Glyph, "\uE14D", "Expected icon glyph to be E14D.");
```
Same here, but we'll swap them around so we test two different chars. 😋

michael-hawker commented on UnitTests/Extensions/Test_FontIconExtensionMarkupExtension.cs:
```diff
@@ -0,0 +1,63 @@
+            Assert.IsNotNull(icon, $"Could not find the {nameof(FontIcon)} element in button.");
+
+            Assert.AreEqual(icon.Glyph, "\uE105", "Expected icon glyph to be E105.");
+            Assert.AreEqual(icon.FontFamily.Source, "Segoe MDL2 Assets", "Expected font family to be Segoe UI");
```
```suggestion
            Assert.AreEqual(icon.FontFamily.Source, "Segoe UI", "Expected font family to be Segoe UI");
```
Appreciate the added tests, I wish VS made test errors easier to read in the test explorer/IDE... 😊

Sergio0694 commented:
@michael-hawker Added the two separate draft PRs, without tests for now:

- `FontIconSource`: #3129 
- `BitmapIcon`: #3130

michael-hawker commented:
Once we're set with the merged feature set, I'll re-test and re-review. I think at that point, we just need to open a Doc PR first in order to flip this over out of Draft state, eh? @Sergio0694

Sergio0694 commented:
@michael-hawker All the other PRs are merged now, sounds good to me! 👍
I can go ahead and put down a draft for the docs PR already in the meantime too!

michael-hawker commented on Microsoft.Toolkit.Uwp.SampleApp/SamplePages/IconExtensions/IconExtensionsPage.xaml:
```diff
@@ -0,0 +1,9 @@
+      xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
+      xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
+      mc:Ignorable="d">
+    <Grid>
```
You should add the markup extensions use here so that they don't get optimized out by the .NET Native compiler in release mode for the sample app, see the readme.

Sergio0694 replied:
Whoops, my bad! Fixed in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3110/commits/bf6a306d0e79b267117f4c182e4c50ce59c918bb, thanks! 😄

michael-hawker replied:
No problem, it's an obscure requirement. 😋

Sergio0694 commented:
@michael-hawker Following up on your suggestion, I've also added some extensions to create icons from `Symbol` values. I've also added one for `FontIconSource`, to get feature parity with the `string`-based extensions. Here are all the new font icon extensions from this PR:

```xml
<CommandBar>

    <!--FontIcon-->
    <AppBarButton Icon="{ex:FontIcon Glyph=&#xE102;}"/>

    <!--SymbolIcon-->
    <AppBarButton Icon="{ex:SymbolIcon Glyph=Play}"/>
</CommandBar>

<SwipeItems>

    <!--FontIconSource-->
    <SwipeItem Text="Accept" IconSource="{ex:FontIconSource Glyph=&#xE10B;}"/>

    <!--SymboliconSource-->
    <SwipeItem Text="Flag" IconSource="{ex:FontIconSource Glyph=Flag}"/>
</SwipeItems>
```

All the icon extensions above also offer extensive customization through their additional properties, such as `Foreground`, `FontSize`, `FontWeight`, `IsTextScaleFactorEnabled`, etc.

Sergio0694 commented:
Ok this last failure is weird, all tests are passing for me locally:

![image](https://user-images.githubusercontent.com/10199417/80019286-93c1ad00-84d7-11ea-93fe-b3d045aee8fb.png)

I wonder if this is some .NET Native issue in Release with the unit test projects 🤔

michael-hawker commented:
@azchohfi @MattWhilden any ideas here on why the rd.xml isn't working? Can we try including the whole assembly for the Unit Test project?

michael-hawker commented on UnitTests/Extensions/Test_BitmapIconExtensionMarkupExtension.cs:
```diff
@@ -0,0 +1,85 @@
+        <Button x:Name=""RootButton"">
+            <Button.Flyout>
+                <MenuFlyout>
+                    <MenuFlyoutItem Icon=""{ex:BitmapIcon Source=/Assets/StoreLogo.png}"" />
```
@Sergio0694 if you use the full classname here, does that work any better in release mode? I wonder if there's an issue going on with the name mapping for some reason?

michael-hawker replied:
Also, can you test if the samples work in Release mode as well with this similar sort of issue or not?

Sergio0694 replied:
As discussed, fixed in https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3110/commits/db20516409a3d507f75a6cf63da488902848dd88 and https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3110/commits/364ea74a6ec2ed3425ed419cc246b520b74b6a81 🎉

michael-hawker commented:
@Sergio0694 this is all set now, eh?

For future reference, we had to add dummy copies of the instances in the Unit Test app's XAML file so that the proper XamlTypeInfo would be generated for use by the XamlReader loading code in the tests.

Sergio0694 commented:
@michael-hawker Yeah this should be good to go as well now! 😊
